### PR TITLE
Implement SmokeTest utilizing Quarkus REST API

### PR DESCRIPTION
**Pull request by ScrewTSW**
### What does this PR do?
* Adds a new smoke test

### Screenshot/screencast of this PR

N/A

### What issues does this PR fix or reference?

#22031

### How to test this PR?

N/A

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [x] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [x] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [x] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [x] [Tests are covering the bugfix or new feature ](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix-or-new-feature)
- [x] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [ ] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### CODE REVIEW

File: tests/e2e/specs/devfiles/SmokeTest.spec.ts [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+const stackName: string = 'Quarkus REST API';
+const projectName: string = 'quarkus-quickstarts';
+const projectAndFileTests: ProjectAndFileTests = e2eContainer.get(CLASSES.ProjectAndFileTests);
+const workspaceHandlingTests: WorkspaceHandlingTests = e2eContainer.get(CLASSES.WorkspaceHandlingTests);
+const loginTests: LoginTests = e2eContainer.get(CLASSES.LoginTests);
+
+suite(`The ${stackName} userstory`, async function () {
+    let projectSection: ViewSection;
+    suite(`Create workspace from ${stackName} simple`, async function () {
```
Comment (dmytro-ndp) [inline tests/e2e/specs/devfiles/SmokeTest.spec.ts]: typo `simple` > `sample`

**Comment by dmytro-ndp** [comment]
@ScrewTSW : target issue https://github.com/eclipse/che/issues/22031 mentioned https://github.com/che-incubator/quarkus-api-example, while this PR is implementing https://github.com/che-incubator/quarkus-api-example

File: tests/e2e/specs/devfiles/SmokeTest.spec.ts (hunk 1 above)
Comment (ScrewTSW) [reply tests/e2e/specs/devfiles/SmokeTest.spec.ts]: fixed

**Comment by nallikaea** [comment]
Is it a good idea to name this test sample spec as SmokeTest.spec.ts not as Quarkus.spec.ts? We can create smoke test suite or smoke test script, which can contain not just one spec file in the future. For example, if we will have another editor and will add smoke for it too.

**Comment by ScrewTSW** [comment]
> Is it a good idea to name this test sample spec as SmokeTest.spec.ts not as Quarkus.spec.ts? We can create smoke test suite or smoke test script, which can contain not just one spec file in the future. For example, if we will have another editor and will add smoke for it too.

Editor configuration should be independent from the test cases. Previously the tests shared the name, but were split into specific folders (theia, che-code)
I don't see this as an issue personally.

The name of the spec is exactly what it does :D It's a smoke test for editor

**Comment by ScrewTSW** [comment]
> @ScrewTSW : target issue #22031 mentioned https://github.com/che-incubator/quarkus-api-example, while this PR is implementing https://github.com/che-samples/quarkus-quickstarts

fixed

File: tests/e2e/tests-library/WorkspaceHandlingTests.ts [hunk 2]
```diff
@@ -66,6 +66,19 @@ export class WorkspaceHandlingTests {
...
 
+    public createAndOpenWorkspaceFromFactory(factoryUrl: string): void {
+        test(`Create and open new workspace from factory:${factoryUrl}`, async () => {
+            await this.dashboard.waitPage();
+            Logger.debug(`Fetching user kubernetes namespace, storing auth token by getting workspaces API URL.`);
+            await this.apiUrlResolver.getWorkspacesApiUrl();
+            await this.dashboard.clickCreateWorkspaceButton();
+            await this.createWorkspace.waitPage();
+            WorkspaceHandlingTests.parentGUID = await this.browserTabsUtil.getCurrentWindowHandle();
+            await this.createWorkspace.startWorkspaceUsingFactory(factoryUrl);
```
Comment (dmytro-ndp) [inline tests/e2e/tests-library/WorkspaceHandlingTests.ts]: There are other steps in test scenario after entering "Create Workspace" page: https://github.com/eclipse/che/issues/22031 : 
> 2. In the `Import from Git` section input https://github.com/che-incubator/quarkus-api-example
> 3. Click `Create & Open` Button

Screenshot from 2023-03-23 12-34-50

Comment (dmytro-ndp) [reply tests/e2e/tests-library/WorkspaceHandlingTests.ts]: Sorry, I misunderstood the code.

The name of method is misleading, IMHO: `startWorkspaceUsingFactory(factoryUrl);` https://github.com/eclipse/che/blob/main/tests/e2e/pageobjects/dashboard/CreateWorkspace.ts#L55

It's literally using section `Import from Git`, which is not obvious.

**Review by dmytro-ndp — APPROVED** [review]
Looks good to merge.

**Comment by ScrewTSW** [comment]
PR updated, new test run exaple:
```
[tdancs@fedora e2e]$ USERSTORY=SmokeTest npm run test-all-devfiles

> @eclipse-che/che-e2e@7.62.0-SNAPSHOT test-all-devfiles
> ./configs/sh-scripts/generateIndex.sh && ./configs/sh-scripts/initDefaultValues.sh && ./configs/sh-scripts/initDevfileTests.sh

Generating index.ts file...
Initialized default values

TS_SELENIUM_VALUE_TLS_SUPPORT =       true
TS_SELENIUM_VALUE_OPENSHIFT_OAUTH =   true
TS_OCP_LOGIN_PAGE_PROVIDER_TITLE =    github-app-sre
E2E_OCP_CLUSTER_VERSION =             4.x
TS_SELENIUM_LOG_LEVEL =               DEBUG
TS_SELENIUM_USERNAME =                REDACTED
TS_SELENIUM_PASSWORD =                REDACTED
TS_SELENIUM_MULTIUSER =               true
TS_SELENIUM_W3C_CHROME_OPTION =       true
NODE_TLS_REJECT_UNAUTHORIZED =        0
TS_SELENIUM_EDITOR =                  che-code

Launching the "SmokeTest" userstory

MOCHA_DIRECTORY = devfiles

> @eclipse-che/che-e2e@7.62.0-SNAPSHOT lint
> tslint --fix -p .

no-unused-variable is deprecated. Since TypeScript 2.9. Please use the built-in compiler checks instead.
no-use-before-declare is deprecated. Since TypeScript 2.9. Please use the built-in compiler checks instead.

> @eclipse-che/che-e2e@7.62.0-SNAPSHOT tsc
> tsc -p .


################## Launch Information ##################

      TS_SELENIUM_BASE_URL: https://che-dogfooding.apps.che-dev.x6e0.p1.openshiftapps.com
      TS_SELENIUM_HEADLESS: false

      TS_SELENIUM_USERNAME: REDACTED
      TS_SELENIUM_PASSWORD: REDACTED

      TS_SELENIUM_EDITOR:   che-code

      TS_SELENIUM_HAPPY_PATH_WORKSPACE_NAME: petclinic-dev-environment
      TS_SELENIUM_DELAY_BETWEEN_SCREENSHOTS: 1000
      TS_SELENIUM_REPORT_FOLDER: ./report
      TS_SELENIUM_EXECUTION_SCREENCAST: false
      DELETE_SCREENCAST_IF_TEST_PASS: true
      TS_SELENIUM_REMOTE_DRIVER_URL: 
      DELETE_WORKSPACE_ON_FAILED_TEST: false
      TS_SELENIUM_LOG_LEVEL: DEBUG
      TS_SELENIUM_LAUNCH_FULLSCREEN: true

      to output timeout variables, set TS_SELENIUM_PRINT_TIMEOUT_VARIABLES to true
 ######################################################## 


  The SmokeTest userstory
    Create workspace from factory:https://github.com/che-incubator/quarkus-api-example.git
          ▼ BrowserTabsUtil.navigateTo https://che-dogfooding.apps.che-dev.x6e0.p1.openshiftapps.com
          ▼ RegularUserOcpCheLoginPage.login
          ▼ OcpRedHatLoginPage.login wait for LogInWithOpenShift page and click button
          ▼ OcpLoginPage.isIdentityProviderLinkVisible
          ▼ OcpLoginPage.clickOnLoginProviderTitle
          ▼ OcpLoginPage.isAuthorizeOpenShiftIdentityProviderPageVisible
          ▼ CheLoginPage.waitFirstBrokerLoginPage
          ▼ OcpLoginPage.isLinkAccountPageVisible
          ▼ TS_SELENIUM_LAUNCH_FULLSCREEN is set to true, maximizing window.
      ✔ Login (17967ms)
          ▼ Dashboard.waitPage
  [ERROR] DriverHelper.waitVisibility - failed with exception, out of attempts - TimeoutError: Waiting for element to be located By(xpath, //div[@id='page-sidebar']//a[contains(text(), 'Workspaces (')])
  Wait timed out after 1071ms
          ▼ Dashboard.waitPage
          ▼ Fetching user kubernetes namespace, storing auth token by getting workspaces API URL.
          ▼ ApiUrlResolver.obtainUserNamespace 
(node:29112) Warning: Setting the NODE_TLS_REJECT_UNAUTHORIZED environment variable to '0' makes TLS connections and HTTPS requests insecure by disabling certificate verification.
(Use `node --trace-warnings ...` to show where the warning was created)
          ▼ ApiUrlResolver.obtainUserNamespace kubeapi success: screwtsw-che-xmlol7
          ▼ Dashboard.clickCreateWorkspaceButton
          ▼ CreateWorkspace.waitPage
          ▼ CreateWorkspace.waitTitleContains text: "Create Workspace"
          ▼ BrowserTabsUtil.getCurrentWindowHandle
          ▼ CreateWorkspace.startWorkspaceUsingFactory factoryUrl: "https://github.com/che-incubator/quarkus-api-example.git"
          ▼ BrowserTabsUtil.waitAndSwitchToAnotherWindow
          ▼ BrowserTabsUtil.getAllWindowHandles
          ▼ BrowserTabsUtil.getAllWindowHandles
          ▼ BrowserTabsUtil.switchToWindow
      ✔ Create and open new workspace from factory:https://github.com/che-incubator/quarkus-api-example.git (7726ms)
  [WARN] WorkspaceHandlingTests.obtainWorkspaceNameFromStartingPage Failed to obtain name from workspace start page, element not visible yet. Retrying.
  [WARN] WorkspaceHandlingTests.obtainWorkspaceNameFromStartingPage Failed to obtain name from workspace start page, element not visible yet. Retrying.
  [WARN] WorkspaceHandlingTests.obtainWorkspaceNameFromStartingPage Failed to obtain name from workspace start page, element not visible yet. Retrying.
  [WARN] WorkspaceHandlingTests.obtainWorkspaceNameFromStartingPage Failed to obtain name from workspace start page, element not visible yet. Retrying.
  [WARN] WorkspaceHandlingTests.obtainWorkspaceNameFromStartingPage Failed to obtain name from workspace start page, element not visible yet. Retrying.
  [WARN] WorkspaceHandlingTests.obtainWorkspaceNameFromStartingPage Failed to obtain name from workspace start page, element not visible yet. Retrying.
  [WARN] WorkspaceHandlingTests.obtainWorkspaceNameFromStartingPage Failed to obtain name from workspace start page, element not visible yet. Retrying.
      • Obtained workspace name from workspace loader page: quarkus-api-example-khcs
      ✔ Obtain workspace name from workspace loader page (7216ms)
          ▼ MochaHooks.registerRunningWorkspace with workspaceName:quarkus-api-example-khcs
      ✔ Register running workspace
      ✔ Wait workspace readiness (19770ms)
          ▼ new SideBarView().getContent().getSection: get quarkus-api-example
      ✔ Check a project folder has been created (88ms)
          ▼ projectSection.findItem: find devfile.yaml
      ✔ Check the project files was imported (3301ms)
          ▼ Dashboard.openDashboard
          ▼ Dashboard.waitPage
          ▼ Dashboard.stopAndRemoveWorkspaceByUI "quarkus-api-example-khcs"
          ▼ Dashboard.stopWorkspaceByUI "quarkus-api-example-khcs"
          ▼ Dashboard.clickWorkspacesButton
          ▼ Workspaces.waitPage
          ▼ Workspaces.waitWorkspaceListItem "quarkus-api-example-khcs"
          ▼ Workspaces.waitWorkspaceWithRunningStatus "quarkus-api-example-khcs"
          ▼ Workspaces.stopWorkspaceByActionsButton
          ▼ Workspaces.waitWorkspaceListItem "quarkus-api-example-khcs"
          ▼ Workspaces.openActionsPopup for the 'quarkus-api-example-khcs' list item
          ▼ Workspaces.clickActionsButton of the 'quarkus-api-example-khcs' list item
          ▼ Workspaces.waitActionsPopup of the 'quarkus-api-example-khcs' list item
          ▼ Workspaces.clickActionsStopWorkspaceButton for the 'quarkus-api-example-khcs' list item
          ▼ Workspaces.waitWorkspaceWithStoppedStatus "quarkus-api-example-khcs"
          ▼ Workspaces.deleteWorkspaceByActionsButton
          ▼ Workspaces.waitWorkspaceListItem "quarkus-api-example-khcs"
          ▼ Workspaces.openActionsPopup for the 'quarkus-api-example-khcs' list item
          ▼ Workspaces.clickActionsButton of the 'quarkus-api-example-khcs' list item
          ▼ Workspaces.waitActionsPopup of the 'quarkus-api-example-khcs' list item
          ▼ Workspaces.clickActionsDeleteButton for the 'quarkus-api-example-khcs' list item
          ▼ Workspaces.waitDeleteWorkspaceConfirmationWindow
          ▼ Workspaces.clickToDeleteConfirmationCheckbox
          ▼ Workspaces.waitEnabledConfirmationWindowDeleteButton
          ▼ Workspaces.waitWorkspaceListItemAbcence "quarkus-api-example-khcs"
      ✔ Stopping and deleting the workspace (13832ms)

      • Chrome driver session stopped.

  8 passing (2m)
  ```

File: tests/e2e/pageobjects/dashboard/CreateWorkspace.ts [hunk 3]
```diff
@@ -67,6 +61,12 @@ export class CreateWorkspace {
         await this.driverHelper.waitAndClick(sampleLocator, timeout);
     }
 
+    async startWorkspaceFromGit(factoryUrl: string, timeout: number = TimeoutConstants.TS_CLICK_DASHBOARD_ITEM_TIMEOUT): Promise<void> {
+        Logger.debug(`CreateWorkspace.startWorkspaceUsingFactory factoryUrl: "${factoryUrl}"`);
```
Comment (dmytro-ndp) [inline tests/e2e/pageobjects/dashboard/CreateWorkspace.ts]: `startWorkspaceUsingFactory` > `startWorkspaceFromGit`

Comment (nallikaea) [reply tests/e2e/pageobjects/dashboard/CreateWorkspace.ts]: startWorkspaceUsingFactory > importFromGitUsingUI

**Comment by nallikaea** [comment]
> > Is it a good idea to name this test sample spec as SmokeTest.spec.ts not as Quarkus.spec.ts? We can create smoke test suite or smoke test script, which can contain not just one spec file in the future. For example, if we will have another editor and will add smoke for it too.
> 
> Editor configuration should be independent from the test cases. Previously the tests shared the name, but were split into specific folders (theia, che-code) I don't see this as an issue personally.
> 
> The name of the spec is exactly what it does :D It's a smoke test for editor

Sorry, I am still thinking that "smoke" should be suite. what do you think about it, @dmytro-ndp, @musienko-maxim ? =)

**Review by ibuziuk — APPROVED** [review]
:+1:

**Comment by nallikaea** [comment]
I suggest another solution: 

`"test-smoke": "./configs/sh-scripts/generateIndex.sh && ./configs/sh-scripts/initDefaultValues.sh npm run lint && npm run tsc && -e USERSTOTY=$Quarkus.spec.js && mocha --config dist/configs/mocharc.js"`

if we will have more then one test specification in our smoke suite we can make changes only in this script in "che", but not in all projects using this script as PR check 

`"test-smoke": "./configs/sh-scripts/generateIndex.sh && ./configs/sh-scripts/initDefaultValues.sh npm run lint && npm run tsc && -e MOCHA_SUITE=$SUITE_NAME && mocha --config dist/configs/mocharc.js"`

**Comment by dmytro-ndp** [comment]
@nallikaea : there will be ONLY ONE SmokeTest.spec.ts per repo.
If we will not link smoke test with particular testing feature name, we will be able to change test scenario in future not fixing test name.

That is why, IMHO, it would better to move `SmokeTest.spec.ts` out of `devfiles` folder, and call the test using the command like following:
> "smoke-test": "./configs/sh-scripts/generateIndex.sh && ./configs/sh-scripts/initDefaultValues.sh npm run lint && npm run tsc && -e MOCHA_DIRECTORY='' -e USERSTORY=SmokeTest && mocha --config dist/configs/mocharc.js"

**Comment by ScrewTSW** [comment]
Currently unable to verify the change against dogfooding
Screenshot_20230329_155058

**Review by dmytro-ndp — APPROVED** [review]
Looks good to merge.

- SmokeTest has been passed: https://main-jenkins-csb-crwqe.apps.ocp-c1.prod.psi.redhat.com/job/Testing/job/e2e/job/basic/job/typescript-tests/7296/console

- EmptyWorkspace has been passed: https://main-jenkins-csb-crwqe.apps.ocp-c1.prod.psi.redhat.com/job/Testing/job/e2e/job/basic/job/typescript-tests/7297/

**Pull request merged**